### PR TITLE
Add date_trunc function to mysql

### PR DESCRIPTION
**Pull request by andresgutgon**
### What is this?
Do a `db_adapter` to abstract `date_trunc`. This way I can implement in Ruby all the methods that generate the necessary methods to do date tuncation on SQL.

### TODO
- [x] Test Mysql migration generator works on a Rails app
- [x] Pass all tests with `DB='mysql' rake test`
- [x] MySQL adapter with `date_trunc`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -142,6 +142,8 @@ ActiveReporting::Configuration.setting = value
 
 `metric_lookup_class` - The name of a constant used to lookup prebuilt `Reporting::Metric` objects by name. The constant should define a class method called `#lookup` which can take a string or symbol of the metric name. (Default: `::Metric`)
 
+`db_adapter` - Each database has a diferent way of doing the same thing on SQL. This is an abstraction layer on each database. For example `PostGreSQL` have a native `date_trunc` while `MySQL` doesn't so we do the same in another way. Make sure to chosse your database's adapter (Default: `:sqlite3`). If you want MySQL put `:mysql2` and if you want PostGreSQL put `:postgresql`.
```
Comment (t27duck) [inline README.md]: Misspelled "choose" here. Also, I've only known the database to be spelled "PostgreSQL". I'm not sure why the "G" is capitalized

Comment (andresgutgon) [reply README.md]: Thanks for the corrections, I'm not an English speaker at all :)

**Comment by andresgutgon** [comment]
Yes, it's finish :). Do you see it ok?

File: README.md [hunk 2]
```diff
@@ -235,7 +237,6 @@ end
 
 When creating a metric, ActiveReporting will recognize implicit hierarchies for this dimension. The hierarchies correspond to the [values](https://www.postgresql.org/docs/8.1/static/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC) supported by PostgreSQL. (See example under the metric section, below.)
 
-*NOTE*: PRs welcomed to support this functionality in other databases.
```
Comment (masylum) [inline README.md]: I guess PRs are still welcome to support other databases 😉

Comment (andresgutgon) [reply README.md]: LOL, true but the rest of databases are SQLitle which nobody use in production. I doubt someone will do that

File: lib/active_reporting/database_adapters/mysql_adapter.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    # Database adapters are here to solve SQL problems in the
+    # most idiomatic way in each database
+    class MysqlAdapter < Base
+      # Generate SQL snippet with DATE_TRUNC
+      # @param [String] interval
+      # @param [String] field
+      # @return [String]
+      def date_trunc(interval, field)
+        clean_sql(
+          <<-SQL
```
Comment (t27duck) [inline lib/active_reporting/database_adapters/mysql_adapter.rb]: The "squiggly heredoc" has been in Ruby since 2.3, I recommend using that instead.
https://ruby-doc.org/core-2.3.0/doc/syntax/literals_rdoc.html#label-Here+Documents

Also, I would say I personally am not concerned by multi-line SQL being generated, so I don't think `#clean_sql` is really necessary.

File: lib/active_reporting/database_adapters/factory.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module ActiveReporting
+  module DatabaseAdapters
+    class Factory
+      class << self
+        ADAPTERS = {
+          sqlite3: SqliteAdapter,
+          mysql2: MysqlAdapter,
```
Comment (t27duck) [inline lib/active_reporting/database_adapters/factory.rb]: Let's go with just `sqlite` and `mysql` here. The use of "3" and "2" in various Rails projects references the adaptor and version of the gem they use. In the case of this gem, we shouldn't care.

File: README.md [hunk 5]
```diff
@@ -142,6 +142,8 @@ ActiveReporting::Configuration.setting = value
 
 `metric_lookup_class` - The name of a constant used to lookup prebuilt `Reporting::Metric` objects by name. The constant should define a class method called `#lookup` which can take a string or symbol of the metric name. (Default: `::Metric`)
 
+`db_adapter` - Each database has a diferent way of doing the same thing on SQL. This is an abstraction layer on each database. For example `PostgreSQL` have a native `date_trunc` while `MySQL` doesn't so we do the same in another way. Make sure to choose your database's adapter (Default: `:sqlite3`). If you want MySQL put `:mysql2` and if you want PostGreSQL put `:postgresql`.
```
Comment (t27duck) [inline README.md]: We should add a note here explaining that while it is encouraged to set this value, it is not required if the user does not with to implement database-specific features such as the date functions.

File: lib/active_reporting/database_adapters/mysql_adapter.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+              "#{super_old_base_date}",
+              INTERVAL TIMESTAMPDIFF(
+                #{interval.upcase},
+                "#{super_old_base_date}",
+                #{field}
+              ) #{interval.upcase}
+            )
+          SQL
+        )
+      end
```
Comment (niborg) [inline lib/active_reporting/database_adapters/mysql_adapter.rb]: I've been playing around with different variants of replicating `date_trunc`, including this one. This approach is about 2x faster according to my primitive metrics, but it falls short in truly replicating `date_trunc`, mostly patently shown by the limit on supported datetime_hierarchies.

Take a look at [this code](https://github.com/LawKick/active_reporting/blob/06a5228728c99d18c4be252362667c8108734af2/lib/active_reporting/function_adapters/mysql.rb#L49) and consider if it would be useful. It is ugly but may provide a better parallel in functionality.

File: test/active_reporting/report_test.rb [hunk 7]
```diff
@@ -32,8 +32,23 @@ def test_report_runs_with_an_aggregate_other_than_count
...
+      data = report.run
+      assert data.all? { |r| r.key?('created_at_week') }
+      assert data.size == 4
+    else
+      assert_raises ActiveReporting::InvalidDimensionLabel do
+        metric = ActiveReporting::Metric.new(:a_metric, fact_model: SaleFactModel, dimensions: [{created_at: :week}])
+        report = ActiveReporting::Report.new(metric)
+      end
+    end
+  end
```
Comment (niborg) [inline test/active_reporting/report_test.rb]: I would propose that you might want to add some more comprehensive tests for the MySQL date_trunc adaptation. You can run into **a lot** of edge cases. See [here](https://github.com/LawKick/active_reporting/blob/master/test/active_reporting/function_adapaters/date_trunc_test.rb).
